### PR TITLE
[Relay/TOPI][OP] Add arange op in Relay and TOPI

### PR DESCRIPTION
**Pull request by icemelon9**
Currently I put start, stop, step in Relay attributes since it is required to infer the output shape. Later if Relay supports unknown dimension like `Any`, we can move them into inputs of arange op instead of attributes.

This PR relies on https://github.com/dmlc/tvm/pull/2615.

### CODE REVIEW

File: include/tvm/relay/attrs/transform.h [hunk 1]
```diff
@@ -96,6 +96,24 @@ struct InitOpAttrs : public tvm::AttrsNode<InitOpAttrs> {
...
+  tvm::Expr stop;
+  tvm::Expr step;
+  DataType dtype;
+
+  TVM_DECLARE_ATTRS(ArangeAttrs, "relay.attrs.ArangeAttrs") {
+    TVM_ATTR_FIELD(start).set_default(make_const(Int(32), 0))
+        .describe("Start of interval. The interval includes this value.");
+    TVM_ATTR_FIELD(stop)
+        .describe("Stop of interval. The interval does not include this value.");
+    TVM_ATTR_FIELD(start).set_default(make_const(Int(32), 1))
```
Comment (zhiics) [inline include/tvm/relay/attrs/transform.h]: s/start/step

File: python/tvm/relay/op/transform.py [hunk 2]
```diff
@@ -249,6 +249,49 @@ def full_like(data, fill_value):
...
+        The resulting tensor.
+
+    Examples
+    --------
+    .. code-block:: python
+
+        relay.arange(5) = [0, 1, 2, 3, 4]
+        relay.arange(1, 5) = [1, 2, 3, 4]
+        relay.arange(1, 5, 1.5) = [1, 2.5, 4]
+    """
```
Comment (zhiics) [inline python/tvm/relay/op/transform.py]: should we consider cases like `start > stop` and `step <= 0`, here? I think we probably need to at least warning or raise exceptions for `step == 0`

File: src/relay/op/tensor/transform.cc [hunk 3]
```diff
@@ -880,6 +880,58 @@ and type as the input array.
...
+TVM_REGISTER_NODE_TYPE(ArangeAttrs);
+
+bool ArangeRel(const Array<Type>& types,
+               int num_inputs,
+               const Attrs& attrs,
+               const TypeReporter& reporter) {
+  CHECK_EQ(types.size(), 1);
+  const ArangeAttrs* param = attrs.as<ArangeAttrs>();
+  IndexExpr num_elem = tvm::cast(tvm::Int(32), tvm::ceil(
+      tvm::cast(tvm::Float(32), param->stop - param->start) / param->step));
```
Comment (zhiics) [inline src/relay/op/tensor/transform.cc]: `CHECK_NE(param->step, 0U)` ?

File: topi/include/topi/transform.h [hunk 4]
```diff
@@ -868,6 +868,19 @@ inline Tensor tensordot(const Tensor& A,
...
 }
 
+inline Tensor arange(const Expr start,
+                     const Expr stop,
+                     const Expr step,
+                     Type dtype,
+                     std::string name = "tensor",
+                     std::string tag = kInjective) {
+  Expr num_elem = tvm::cast(tvm::Int(32), tvm::ceil(
+      tvm::cast(tvm::Float(32), stop - start) / step));
```
Comment (zhiics) [inline topi/include/topi/transform.h]: not sure if we need to check if `step == 0`, probably it is enough if we checked if before

Comment (icemelon9) [reply topi/include/topi/transform.h]: divide by 0 should be captured by IR when step is constant.
https://github.com/dmlc/tvm/blob/master/src/lang/ir_operator.cc#L202

File: src/relay/op/tensor/transform.cc (hunk 3 above)
Comment (icemelon9) [reply src/relay/op/tensor/transform.cc]: `step` is not necessary to be constant during the compilation time. So probably we should rely on IR to capture this?

File: include/tvm/relay/attrs/transform.h (hunk 1 above)
Comment (icemelon9) [reply include/tvm/relay/attrs/transform.h]: Fixed.

File: src/relay/op/tensor/transform.cc (hunk 3 above)
Comment (zhiics) [reply src/relay/op/tensor/transform.cc]: I see. Thanks.

**Comment by Laurawly** [comment]
@zhreshold

File: python/tvm/relay/op/transform.py (hunk 2 above)
Comment (icemelon9) [reply python/tvm/relay/op/transform.py]: Added the sanity check in the new commit

**Comment by icemelon9** [comment]
Hey @yzhliu, could you help review this PR?

File: tests/python/relay/test_op_level3.py [hunk 5]
```diff
@@ -457,6 +457,40 @@ def test_infer_type_prelu():
     verify_infer_type_prelu((1, 3, 2, 2), None, 1, (1, 3, 2, 2))
     verify_infer_type_prelu((1, 2, 2, 3), None, 3, (1, 2, 2, 3))
 
+
+def test_arange():
```
Comment (zhreshold) [inline tests/python/relay/test_op_level3.py]: can we also have a relay frontend test from mxnet arange

Comment (icemelon9) [reply tests/python/relay/test_op_level3.py]: Sure. will add.

File: src/relay/op/tensor/transform.cc [hunk 6]
```diff
@@ -880,6 +880,61 @@ and type as the input array.
...
+  if (const tvm::ir::IntImm* val = num_elem.as<tvm::ir::IntImm>()) {
+    CHECK_GT(val->value, 0) << "Invalid arange inputs";
```
Comment (yzhliu) [inline src/relay/op/tensor/transform.cc]: suggest to also print related params

Comment (icemelon9) [reply src/relay/op/tensor/transform.cc]: Fixed in the new commit

File: python/tvm/relay/op/transform.py [hunk 7]
```diff
@@ -249,6 +249,49 @@ def full_like(data, fill_value):
     return _make.full_like(data, fill_value)
 
 
+def arange(stop, start=None, step=1, dtype="float32"):
```
[the remark this answers is not in the document]
Comment (icemelon9) [reply python/tvm/relay/op/transform.py]: Made a note in docs in the new commit

File: tests/python/relay/test_op_level3.py (hunk 5 above)
Comment (icemelon9) [reply tests/python/relay/test_op_level3.py]: Added in the new commit.

**Comment by yzhliu** [comment]
Thanks all. This is merged.

**Pull request merged**